### ISSUE TITLE
Pothos-like error handling

### ISSUE THREAD

**Issue by satvikpendem** [async-graphql/async-graphql#1267]
I used [Pothos](https://pothos-graphql.dev/) on the TypeScript/Node.js side before and I really liked their [error handling](https://pothos-graphql.dev/docs/plugins/errors), which, for some query `hello`, basically created a schema that looks like this:

```graphql
type Error {
  message: String!
}

type Query {
  hello(name: String!): QueryHelloResult
}

union QueryHelloResult = Error | QueryHelloSuccess

type QueryHelloSuccess {
  data: String!
}
```

and could be queried like this:

```graphql
query {
  hello(name: "World") {
    __typename
    ... on Error {
      message
    }
    ... on QueryHelloSuccess {
      data
    }
  }
}
```

This basically forced the user to handle both the success case and the error case, or the query would fail. Is this something we could add for async-graphql too, particularly the part about having the user having to handle all cases in the query itself rather than detecting whether an `errors` field exists?

**Comment by satvikpendem** [comment]
Any thoughts on this style of error handling @sunli829? Personally I think it could be good to enforce error side type safety like this just like how Rust's Result type handles it, just on the client side. Pothos uses an error plugin that generates the necessary unions, perhaps we could use a macro to do this as well.

**Comment by gusinacio** [comment]
I really like this idea. Our current frontend is struggling with error handling because the errors aren't typed.

It would be nice if we have some way to enable unions within Results, basically a `Result<Object, Error>` could become an `Union` or a "flatten" `Union` = Object | Error

**Comment by satvikpendem** [comment]
@gusinacio Indeed, currently I'm just generating all of the unions manually but it would be great to have an extension in `async-graphql` directly such as with a macro annotation.

**Comment by musjj** [comment]
I'm also struggling with this, since `async_graphql` only allows unit enums. Is there a reason behind this limitation? Right now enum-based results are not even possible due to this limitation:

```rust
enum QueryResult {
    InternalError,
    Success { value: String },
}
```

Is there a workaround to this?
